plugins/oauth/server/providers/base.py

```diff
@@ -0,0 +1,173 @@
+
+    def getUser(self, code):
+        """
+        Returns a user via this OAuth provider given an OAuth code. Subclasses
+        must implement this to retrieve an email address, first name, last name,
+        and optionally a login name from the OAuth provider, and then call
+        self.createOrReuseUser with that information to lookup or create a
+        corresponding Girder user.
+        """
+        raise NotImplementedError()
+
+    def createOrReuseUser(self, email, firstName, lastName, userName=None):
+        cursor = self.model('user').find({'email': email}, limit=1)
+        if cursor.count(True) == 0:
+            policy = self.model('setting').get(SettingKey.REGISTRATION_POLICY)
+            if policy != 'open':
+                raise RestException(
+                    'Registration on this instance is closed. Contact an '
+                    'administrator to create an account for you.')
+            login = self.deriveLogin(email, firstName, lastName, userName)
+
+            user = self.model('user').createUser(
+                login=login, password=None, firstName=firstName,
+                lastName=lastName, email=email)
+        else:
+            user = cursor[0]
+            user['firstName'] = firstName
+            user['lastName'] = lastName
+
+        return user
+
+    def generateLogins(self, email, firstName, lastName, userName=None):
+        """
+        Generate a series of reasonable login names for a new user based on
+        their basic information sent to us by the provider.
+        """
+        # If they have a username on the other service, try that
+        if userName:
+            yield userName
+            userName = re.sub('[\W_]+', '', userName)
+            yield userName
+
+            for i in range(1, 6):
+                yield '%s%d' % (userName, i)
+
+        # Next try to use the prefix from their email address
+        prefix = email.split('@')[0]
+        yield prefix
+        yield re.sub('[\W_]+', '', prefix)
+
+        # Finally try to use their first and last name
+        yield '%s%s' % (firstName, lastName)
+
+        for i in range(1, 6):
+            yield '%s%s%d' % (firstName, lastName, i)
+
+    def deriveLogin(self, email, firstName, lastName, userName=None):
+        """
+        Attempt to automatically create a login name from existing user
+        information from OAuth providers. Attempts to generate it from the
+        username on the provider, the email address, or first and last name. If
+        not possible, returns None and it is left to the caller to generate
+        their own login for the user or choose to fail.
+
+        :param email: The email address.
+        :type email: str
+        """
+        for login in self.generateLogins(email, firstName, lastName, userName):
+            login = login.lower()
+            if self.testLogin(login):
+                return login
+
+        return None
```

Comment:
An exception should be raised at this point. If we absolutely can't create a login name, we shouldn't return the invalid login name ``None``, since that will just fail later in the more-difficult-to-trace ``User.validate`` method.